File: python_komega/komega_math.py
"""
Komega Math Module - Mathematical operations and BLAS/LAPACK wrappers

This module corresponds to the Fortran komega_math module and provides
mathematical operations with MPI support for distributed computing.

Copyright (C) 2016 Mitsuaki Kawamura
Python port created for verification and testing purposes.
"""

from typing import Optional, Union

import numpy as np

from komega_parameter import get_global_params


class KomegaMath:
    """
    Mathematical operations for the Komega library.

    This class provides BLAS/LAPACK-like operations with MPI support
    for distributed computing environments.
    """

    def __init__(self):
        """Initialize the math operations class."""
        self.params = get_global_params()

    def ddot(self, x: np.ndarray, y: np.ndarray) -> float:
        """
        Compute dot product of two real vectors.

        Parameters
        ----------
        x : np.ndarray
            First vector
        y : np.ndarray
            Second vector

        Returns
        -------
        float
            Dot product of x and y
        """
        return np.dot(x, y)

    def zdotc(self, x: np.ndarray, y: np.ndarray) -> complex:
        """
        Compute conjugate dot product of two complex vectors.

        Parameters
        ----------
        x : np.ndarray
            First vector
        y : np.ndarray
            Second vector

        Returns
        -------
        complex
            Conjugate dot product of x and y
        """
        return np.dot(np.conj(x), y)

    def zdotu(self, x: np.ndarray, y: np.ndarray) -> complex:
        """
        Compute dot product of two complex vectors.

        Parameters
        ----------
        x : np.ndarray
            First vector
        y : np.ndarray
            Second vector

        Returns
        -------
        complex
            Dot product of x and y
        """
        return np.dot(x, y)

    def dscal(self, alpha: float, x: np.ndarray) -> None:
        """
        Scale a real vector by a scalar.

        Parameters
        ----------
        alpha : float
            Scaling factor
        x : np.ndarray
            Vector to scale (modified in place)
        """
        x *= alpha

    def zscal(self, alpha: complex, x: np.ndarray) -> None:
        """
        Scale a complex vector by a scalar.

        Parameters
        ----------
        alpha : complex
            Scaling factor
        x : np.ndarray
            Vector to scale (modified in place)
        """
        x *= alpha

    def dcopy(self, x: np.ndarray, y: np.ndarray) -> None:
        """
        Copy a real vector.

        Parameters
        ----------
        x : np.ndarray
            Source vector
        y : np.ndarray
            Destination vector (modified in place)
        """
        y[:] = x[:]

    def zcopy(self, x: np.ndarray, y: np.ndarray) -> None:
        """
        Copy a complex vector.

        Parameters
        ----------
        x : np.ndarray
            Source vector
        y : np.ndarray
            Destination vector (modified in place)
        """
        y[:] = x[:]

    def daxpy(self, alpha: float, x: np.ndarray, y: np.ndarray) -> None:
        """
        Compute y = alpha * x + y for real vectors.

        Parameters
        ----------
        alpha : float
            Scaling factor
        x : np.ndarray
            First vector
        y : np.ndarray
            Second vector (modified in place)
        """
        y += alpha * x

    def zaxpy(self, alpha: complex, x: np.ndarray, y: np.ndarray) -> None:
        """
        Compute y = alpha * x + y for complex vectors.

        Parameters
        ----------
        alpha : complex
            Scaling factor
        x : np.ndarray
            First vector
        y : np.ndarray
            Second vector (modified in place)
        """
        y += alpha * x

    def ddotMPI(self, x: np.ndarray, y: np.ndarray) -> float:
        """
        Compute dot product with MPI reduction.

        Parameters
        ----------
        x : np.ndarray
            First vector
        y : np.ndarray
            Second vector

        Returns
        -------
        float
            Dot product with MPI reduction
        """
        result = self.ddot(x, y)

        # In MPI version, this would perform MPI_Allreduce
        if self.params.lmpi:
            # For now, just return the local result
            # In a real MPI implementation, this would call MPI_Allreduce
            pass

        return result

    def zdotcMPI(self, x: np.ndarray, y: np.ndarray) -> complex:
        """
        Compute conjugate dot product with MPI reduction.

        Parameters
        ----------
        x : np.ndarray
            First vector
        y : np.ndarray
            Second vector

        Returns
        -------
        complex
            Conjugate dot product with MPI reduction
        """
        result = self.zdotc(x, y)

        # In MPI version, this would perform MPI_Allreduce
        if self.params.lmpi:
            # For now, just return the local result
            # In a real MPI implementation, this would call MPI_Allreduce
            pass

        return result

    def zdotuMPI(self, x: np.ndarray, y: np.ndarray) -> complex:
        """
        Compute dot product with MPI reduction.

        Parameters
        ----------
        x : np.ndarray
            First vector
        y : np.ndarray
            Second vector

        Returns
        -------
        complex
            Dot product with MPI reduction
        """
        result = self.zdotu(x, y)

        # In MPI version, this would perform MPI_Allreduce
        if self.params.lmpi:
            # For now, just return the local result
            # In a real MPI implementation, this would call MPI_Allreduce
            pass

        return result

    def dabsmax(self, array: np.ndarray) -> float:
        """
        Find maximum absolute value with MPI reduction.

        Parameters
        ----------
        array : np.ndarray
            Input array

        Returns
        -------
        float
            Maximum absolute value
        """
        result: float = np.max(np.abs(array))

        # In MPI version, this would perform MPI_Allreduce with MPI_MAX
        if self.params.lmpi:
            # For now, just return the local result
            # In a real MPI implementation, this would call MPI_Allreduce
            pass

        return result

    def norm(self, x: np.ndarray) -> float:
        """
        Compute the Euclidean norm of a vector.

        Parameters
        ----------
        x : np.ndarray
            Input vector

        Returns
        -------
        float
            Euclidean norm
        """
        return np.sqrt(np.real(self.zdotcMPI(x, x)))

    def complex_norm(self, x: np.ndarray) -> float:
        """
        Compute the Euclidean norm of a complex vector.

        Parameters
        ----------
        x : np.ndarray
            Input complex vector

        Returns
        -------
        float
            Euclidean norm
        """
        return np.sqrt(np.real(self.zdotcMPI(x, x)))


# Global math operations instance
_global_math = KomegaMath()


def get_global_math() -> KomegaMath:
    """
    Get the global math operations instance.

    Returns
    -------
    KomegaMath
        Global math operations instance
    """
    return _global_math


# Convenience functions for direct access
def ddot(x: np.ndarray, y: np.ndarray) -> float:
    """Compute dot product of two real vectors."""
    return _global_math.ddot(x, y)


def zdotc(x: np.ndarray, y: np.ndarray) -> complex:
    """Compute conjugate dot product of two complex vectors."""
    return _global_math.zdotc(x, y)


def zdotu(x: np.ndarray, y: np.ndarray) -> complex:
    """Compute dot product of two complex vectors."""
    return _global_math.zdotu(x, y)


def dscal(alpha: float, x: np.ndarray) -> None:
    """Scale a real vector by a scalar."""
    _global_math.dscal(alpha, x)


def zscal(alpha: complex, x: np.ndarray) -> None:
    """Scale a complex vector by a scalar."""
    _global_math.zscal(alpha, x)


def dcopy(x: np.ndarray, y: np.ndarray) -> None:
    """Copy a real vector."""
    _global_math.dcopy(x, y)


def zcopy(x: np.ndarray, y: np.ndarray) -> None:
    """Copy a complex vector."""
    _global_math.zcopy(x, y)


def daxpy(alpha: float, x: np.ndarray, y: np.ndarray) -> None:
    """Compute y = alpha * x + y for real vectors."""
    _global_math.daxpy(alpha, x, y)


def zaxpy(alpha: complex, x: np.ndarray, y: np.ndarray) -> None:
    """Compute y = alpha * x + y for complex vectors."""
    _global_math.zaxpy(alpha, x, y)


def ddotMPI(x: np.ndarray, y: np.ndarray) -> float:
    """Compute dot product with MPI reduction."""
    return _global_math.ddotMPI(x, y)


def zdotcMPI(x: np.ndarray, y: np.ndarray) -> complex:
    """Compute conjugate dot product with MPI reduction."""
    return _global_math.zdotcMPI(x, y)


def zdotuMPI(x: np.ndarray, y: np.ndarray) -> complex:
    """Compute dot product with MPI reduction."""
    return _global_math.zdotuMPI(x, y)


def dabsmax(array: np.ndarray) -> float:
    """Find maximum absolute value with MPI reduction."""
    return _global_math.dabsmax(array)

File: c_komega/src/komega_math.c
/**
 * @file komega_math.c
 * @brief Mathematical operations module for Komega C library
 * 
 * This module provides mathematical operations similar to BLAS functions.
 * It corresponds to the komega_math Fortran module and includes
 * dot products, scaling, axpy operations, and other mathematical utilities.
 * 
 * @author Mitsuaki Kawamura (Original Fortran)
 * @author C Port for verification and testing purposes
 * @date 2025
 * @version 1.0.0
 * 
 * @see komega_math.h
 * @see komega.h
 */

#include "komega_math.h"
#include <math.h>
#include <stdlib.h>

/* Global math instance */
static komega_math_t* global_math = NULL;

/**
 * @brief Create a new math instance
 * 
 * Allocates memory for a new math structure and initializes it.
 * 
 * @return Pointer to new math instance, or NULL on failure
 * 
 * @see komega_math_destroy()
 */
komega_math_t* komega_math_create(void) {
    komega_math_t* math = malloc(sizeof(komega_math_t));
    if (math == NULL) {
        return NULL;
    }
    
    /* Initialize with default values */
    math->initialized = true;
    
    return math;
}

/**
 * @brief Destroy a math instance
 * 
 * Frees the memory allocated for the math structure.
 * Safe to call with NULL pointer.
 * 
 * @param math Pointer to math instance to destroy
 * 
 * @see komega_math_create()
 */
void komega_math_destroy(komega_math_t* math) {
    if (math != NULL) {
        free(math);
    }
}

/**
 * @brief Real dot product
 * 
 * Computes the dot product of two real vectors: result = x^T * y
 * 
 * @param x First vector
 * @param y Second vector
 * @param n Vector length
 * @return Dot product result
 */
double komega_math_ddot(const double* x, const double* y, int n) {
    double result = 0.0;
    for (int i = 0; i < n; i++) {
        result += x[i] * y[i];
    }
    return result;
}

/**
 * @brief Complex conjugate dot product
 * 
 * Computes the conjugate dot product of two complex vectors: result = x^H * y
 * 
 * @param x First vector
 * @param y Second vector
 * @param n Vector length
 * @return Conjugate dot product result
 */
komega_complex_t komega_math_zdotc(const komega_complex_t* x, 
                                   const komega_complex_t* y, int n) {
    komega_complex_t result = 0.0 + 0.0 * I;
    for (int i = 0; i < n; i++) {
        result += conj(x[i]) * y[i];
    }
    return result;
}

/**
 * @brief Complex dot product
 * 
 * Computes the dot product of two complex vectors: result = x^T * y
 * 
 * @param x First vector
 * @param y Second vector
 * @param n Vector length
 * @return Dot product result
 */
komega_complex_t komega_math_zdotu(const komega_complex_t* x, 
                                   const komega_complex_t* y, int n) {
    komega_complex_t result = 0.0 + 0.0 * I;
    for (int i = 0; i < n; i++) {
        result += x[i] * y[i];
    }
    return result;
}

/**
 * @brief Real vector scaling
 * 
 * Scales a real vector by a scalar: x = alpha * x
 * 
 * @param alpha Scaling factor
 * @param x Vector to scale (modified in place)
 * @param n Vector length
 */
void komega_math_dscal(double alpha, double* x, int n) {
    for (int i = 0; i < n; i++) {
        x[i] *= alpha;
    }
}

/**
 * @brief Complex vector scaling
 * 
 * Scales a complex vector by a scalar: x = alpha * x
 * 
 * @param alpha Scaling factor
 * @param x Vector to scale (modified in place)
 * @param n Vector length
 */
void komega_math_zscal(komega_complex_t alpha, komega_complex_t* x, int n) {
    for (int i = 0; i < n; i++) {
        x[i] *= alpha;
    }
}

/**
 * @brief Real axpy operation
 * 
 * Performs y = alpha * x + y
 * 
 * @param alpha Scaling factor
 * @param x First vector
 * @param y Second vector (modified in place)
 * @param n Vector length
 */
void komega_math_daxpy(double alpha, const double* x, double* y, int n) {
    for (int i = 0; i < n; i++) {
        y[i] += alpha * x[i];
    }
}

/**
 * @brief Complex axpy operation
 * 
 * Performs y = alpha * x + y
 * 
 * @param alpha Scaling factor
 * @param x First vector
 * @param y Second vector (modified in place)
 * @param n Vector length
 */
void komega_math_zaxpy(komega_complex_t alpha, const komega_complex_t* x, 
                       komega_complex_t* y, int n) {
    for (int i = 0; i < n; i++) {
        y[i] += alpha * x[i];
    }
}

/**
 * @brief Real vector copy
 * 
 * Copies a real vector: y = x
 * 
 * @param x Source vector
 * @param y Destination vector
 * @param n Vector length
 */
void komega_math_dcopy(const double* x, double* y, int n) {
    for (int i = 0; i < n; i++) {
        y[i] = x[i];
    }
}

/**
 * @brief Complex vector copy
 * 
 * Copies a complex vector: y = x
 * 
 * @param x Source vector
 * @param y Destination vector
 * @param n Vector length
 */
void komega_math_zcopy(const komega_complex_t* x, komega_complex_t* y, int n) {
    for (int i = 0; i < n; i++) {
        y[i] = x[i];
    }
}

/**
 * @brief Real vector absolute maximum
 * 
 * Finds the maximum absolute value in a real vector
 * 
 * @param x Vector
 * @param n Vector length
 * @return Maximum absolute value
 */
double komega_math_dabsmax(const double* x, int n) {
    double max_val = 0.0;
    for (int i = 0; i < n; i++) {
        double abs_val = fabs(x[i]);
        if (abs_val > max_val) {
            max_val = abs_val;
        }
    }
    return max_val;
}

/**
 * @brief Complex vector absolute maximum
 * 
 * Finds the maximum absolute value in a complex vector
 * 
 * @param x Vector
 * @param n Vector length
 * @return Maximum absolute value
 */
double komega_math_zabsmax(const komega_complex_t* x, int n) {
    double max_val = 0.0;
    for (int i = 0; i < n; i++) {
        double abs_val = cabs(x[i]);
        if (abs_val > max_val) {
            max_val = abs_val;
        }
    }
    return max_val;
}

/**
 * @brief Real dot product (MPI-like interface)
 * 
 * MPI-like interface for real dot product computation.
 * Currently simulated for single-threaded execution.
 * 
 * @param x First vector
 * @param y Second vector
 * @param n Vector length
 * @return Dot product result
 */
double komega_math_ddot_mpi(const double* x, const double* y, int n) {
    return komega_math_ddot(x, y, n);
}

/**
 * @brief Complex conjugate dot product (MPI-like interface)
 * 
 * MPI-like interface for complex conjugate dot product computation.
 * Currently simulated for single-threaded execution.
 * 
 * @param x First vector
 * @param y Second vector
 * @param n Vector length
 * @return Conjugate dot product result
 */
komega_complex_t komega_math_zdotc_mpi(const komega_complex_t* x, 
                                       const komega_complex_t* y, int n) {
    return komega_math_zdotc(x, y, n);
}

/**
 * @brief Complex dot product (MPI-like interface)
 * 
 * MPI-like interface for complex dot product computation.
 * Currently simulated for single-threaded execution.
 * 
 * @param x First vector
 * @param y Second vector
 * @param n Vector length
 * @return Dot product result
 */
komega_complex_t komega_math_zdotu_mpi(const komega_complex_t* x, 
                                       const komega_complex_t* y, int n) {
    return komega_math_zdotu(x, y, n);
}

/**
 * @brief Get the global math instance
 * 
 * Returns a singleton instance of the math structure.
 * Creates a new instance if none exists.
 * 
 * @return Pointer to global math instance
 * 
 * @see komega_cleanup_global_math()
 */
komega_math_t* komega_get_global_math(void) {
    if (global_math == NULL) {
        global_math = komega_math_create();
    }
    return global_math;
}

/**
 * @brief Clean up the global math instance
 * 
 * Destroys the global math instance and frees its memory.
 * 
 * @see komega_get_global_math()
 */
void komega_cleanup_global_math(void) {
    if (global_math != NULL) {
        komega_math_destroy(global_math);
        global_math = NULL;
    }
}
